### ISSUE TITLE
Suggestion: Minimize converting between strings and date/time

### ISSUE THREAD

**Issue by raman325** [gcobb321/icloud3#68]
Hi Gary,

I noticed that you are doing a lot of manipulation with date/times and strings and converting data back and forth throughout the code. Is there a particular reason for this? I'm wondering if you'd be open to refactoring that code to preserve datetimes and/or times as much as possible in the native format? You'd then get native Python benefits that would simplify the code, and I think you could hand off timezone handling to HA. I'd be open to giving this a shot, but only if you think it would be worthwhile.

**Comment by gcobb321** [comment]
1. The date/times are in many different formats in HA itself, in the iosapp timestamp, the Find-my-Phone timestamp, the Find-my-Friends timestamp, iC3 basically cycles in eseconds. The various routines convert the different timestamp formats into eseconds to determine the polling interval, next update time, last update time, last located time, last change time, etc. In some cases, the time is also kept as a string value in a table so they do not have to be converted back and forth when used in the device_tracker fields, sensors states for the attribute, log files, event logs, etc. I'm sure there is some tuning and recoding that will make them more efficient. Give it a shot.

2. I saw that you have issued several pull requests. The one for the waze/calc parameter looks like another approach to verifying the variables and might be a little more efficient than the way I did it. The approach I took when loading with bad waze regions was to continue to load, assume the default value and issue a warning rather than aborting the load process. Yours is to abort with an error message. I'll give it some thought.

3. Another pull request was reformatting the code using 'Black' (?). I find that having variables with '=' signs that all line up makes the code much more readable. With the reformatting, all internal spaces are removed and I find that when the code is over 6500 lines long, it is much more difficult to review and debug the code. So I will not be using that request.

4. The pull requests you have issued are all against production code. Since I am the only one working on iCloud3, I have set up a development environment on my computer and pi for development, debugging, etc. I am not using github for it but have found that keeping multiple copies and backups of code as changes as they are verified works best for me. Only after I have tested everything, including regression testing to make sure I have not introduced unexpected problems, do I promote it to the beta section of the iCloud3 github repository. When I commit it to the repository, it is then available for usage. I am getting ready to put up a second repository, 'icloud3_dev' to be able to check out some changes and new documentation before it is promoted to the icloud3 production repository. 

Also, v1.0.6 is about ready to be replaced with v1.1.0 (currently in beta 8 but beta 9 is about finished). The release candidate (beta 9) will be uploaded to the icloud3_dev repository. If you want to contribute, which is fine, any changes will need to be made to this development version. That will give me the opportunity to review any changes and possibly roll them into the next update. Normal updates provide new function. Updates that are just code cleanup and optimization are rolled when new features are available, unless they are significant.

Thanks for your interest and I welcome your contribution and ideas.

Gary

**Comment by raman325** [comment]
Gary,

Thanks for the comprehensive response. One by one:

1. I suspected that was the case about different timestamp formats because of different sources but I didn't spend much time looking in the code to validate that. It may still make sense to have a universal data type for your date/time formats, especially as you run different calculations or comparisons against two timestamps, but that may not be prove to be correct. Anyway, I did see that the new version is coming so I will wait until you've had time to fully develop it before I attempt to implement a change like thi.

2. Oops, the PR I submitted was supposed to be just for the README update, but I was looking at this late at night and forgot that when I committed changes to my fork after submitting the PR, that it would update the upstream PR as well. I will revert this, as I planned to submit any additional changes as separate PRs to make it easier for you to review. I did feel my changes were a bit intrusive so I wasn't ready to suggest them yet.

3. See #2. This was not meant to be a suggested change actually, I just was trying to restructure some of the code for myself so that I could play around with it. [Black](https://black.readthedocs.io/en/stable/) is an automatic code formatting tool. The HomeAssistant team recently decided to start using it for all of their code so I have been experimenting with it.

4. Yes this workflow makes sense based on the way your repo is set up. I wasn't clear on the state of the beta code so all of my efforts to try to understand the logic were spent on the "production" instance, hence the accidental commits included in the PR.

Thanks!
